pkg/controller/direct/apigee/instance_controller.go

```diff
@@ -0,0 +1,264 @@
+	operationsClient *api.OrganizationsOperationsService
+	desired          *krm.ApigeeInstance
+	actual           *api.GoogleCloudApigeeV1Instance
+}
+
+var _ directbase.Adapter = &ApigeeInstanceAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *ApigeeInstanceAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting ApigeeInstance", "name", a.id)
+
+	googlecloudapigeev1instancepb, err := a.instancesClient.Get(a.id.String()).Context(ctx).Do()
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ApigeeInstance %q: %w", a.id, err)
+	}
+
+	a.actual = googlecloudapigeev1instancepb
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *ApigeeInstanceAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating ApigeeInstance", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := ApigeeInstanceSpec_ToAPI(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Name = a.id.ID()
+
+	op, err := a.instancesClient.Create(a.id.Parent().String(), resource).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("creating ApigeeInstance %s: %w", a.id, err)
+	}
+	if err := a.waitForOp(ctx, op); err != nil {
+		return fmt.Errorf("ApigeeInstance %s waiting creation: %w", a.id, err)
+	}
+
+	created, err := a.instancesClient.Get(a.id.String()).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("getting created ApigeeInstance: %w", err)
+	}
+
+	log.V(2).Info("successfully created ApigeeInstance", "name", a.id)
+
+	status := &krm.ApigeeInstanceStatus{}
+	status.ObservedState = ApigeeInstanceObservedState_FromAPI(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *ApigeeInstanceAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating ApigeeInstance", "name", a.id)
+	mapCtx := &direct.MapContext{}
+	updateMask := fieldmaskpb.FieldMask{}
+
+	resource := ApigeeInstanceSpec_ToAPI(mapCtx, &a.desired.DeepCopy().Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	if resource.AccessLoggingConfig != nil && !reflect.DeepEqual(resource.AccessLoggingConfig, a.actual.AccessLoggingConfig) {
```

Comment:
Can we use this function to sort and the compare? https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/direct/apigee/envgroup_controller.go#L278